src/core/regimes.rkt

```diff
@@ -25,16 +25,33 @@
 ;; `infer-splitpoints` and `combine-alts` are split so the mainloop
 ;; can insert a timeline break between them.
 
-(define (infer-splitpoints alts ctx)
-  (timeline-event! 'regimes)
+(define (infer-splitpoints alts branch-exprs cerrs try-first ctx)
   (timeline-push! 'inputs (map (compose ~a alt-expr) alts))
-  (define branch-exprs
-    (if (flag-set? 'reduce 'branch-expressions)
-        (exprs-to-branch-on alts ctx)
-        (context-vars ctx)))
   (define err-lsts (batch-errors (map alt-expr alts) (*pcontext*) ctx))
-  (define options (for/list ([bexpr branch-exprs]) (option-on-expr alts err-lsts bexpr ctx)))
-  (define best (argmin (compose errors-score option-errors) options))
+
+  ;; we want to try the branch expression that was best in the previous iteration first
+  ;; and then if there are other branch-exprs that are more accurate on a larger set of alts, try it.
+  (define init (option-on-expr alts err-lsts try-first ctx)) 
+  (define init-err (+ (errors-score (option-errors init)) (length (option-split-indices init))))
+
+  ;; invariant:
+  ;; errs[bexpr] is some best option on branch expression bexpr computed on more alts than we have right now.
+  (define-values (best best-err best-branch errs) 
+      (for/fold ([best init] [best-err init-err] [best-branch try-first] [errs (hash-set cerrs try-first init-err)] 
+            #:result (values best best-err best-branch errs)) 
+            ([bexpr branch-exprs])
+    ;; don't recompute if we know we've computed this branch-expr on more alts and it's still worse
+    (cond 
+      [(and (< (hash-ref cerrs bexpr) best-err) (not (equal? bexpr try-first)))
+        (define opt (option-on-expr alts err-lsts bexpr ctx))
```

Comment:
Would we perhaps be interested in adding the number of option evaluations to the timeline? That seems like an interesting metric and something you've probably been tracking.

Reply:
Actually, I just checked and `option-on-expr` already does this.